system-addon/test/unit/content-src/components/TopSites.test.jsx

```diff
@@ -61,51 +61,73 @@ describe("<TopSitesPerfTimer>", () => {
 
       assert.calledOnce(stub);
       assert.calledWithExactly(stub, instance._sendPaintedEvent);
     });
-    it("should not call _onNextFrame if props.TopSites.initialized is false", () => {
+    it("should not call _afterFramePaint if props.TopSites.initialized is false", () => {
       sandbox.stub(DEFAULT_PROPS.TopSites, "initialized").value(false);
       const wrapper = shallow(<TopSitesPerfTimer {...DEFAULT_PROPS} />);
       const instance = wrapper.instance();
-      const stub = sandbox.stub(instance, "_onNextFrame");
+      const stub = sandbox.stub(instance, "_afterFramePaint");
 
       instance._maybeSendPaintedEvent();
 
       assert.notCalled(stub);
     });
-    it("should not call _onNextFrame if this._timestampSent is true", () => {
+
+    it("should not call _afterFramePaint if this._timestampHandled is true", () => {
       const wrapper = shallow(<TopSitesPerfTimer {...DEFAULT_PROPS} />);
       const instance = wrapper.instance();
-      const stub = sandbox.stub(instance, "_onNextFrame");
-      instance._timestampSent = true;
+      const stub = sandbox.stub(instance, "_afterFramePaint");
+      instance._timestampHandled = true;
 
       instance._maybeSendPaintedEvent();
 
       assert.notCalled(stub);
     });
-  });
 
-  describe("#_onNextFrame", () => {
-    it("should call callback one frame after the current one", () => {
-      const callback = sandbox.spy();
+    it("should set this._timestampHandled=true when called with Topsites.initialized === true", () => {
       const wrapper = shallow(<TopSitesPerfTimer {...DEFAULT_PROPS} />);
+      const instance = wrapper.instance();
+      sandbox.stub(instance, "_afterFramePaint");
+      instance._timestampHandled = false;
 
+      instance._maybeSendPaintedEvent();
+
+      assert.isTrue(instance._timestampHandled);
+    });
+    it("should not set this._timestampHandled=true when called with Topsites.initialized === false", () => {
+      let props = {};
+      Object.assign(props, DEFAULT_PROPS, {TopSites: {initialized: false}});
+      const wrapper = shallow(<TopSitesPerfTimer {...props} />);
       const instance = wrapper.instance();
-      instance._onNextFrame(callback);
+      sandbox.stub(instance, "_afterFramePaint");
+      instance._timestampHandled = false;
 
-      mockRaf.step({count: 1});
-      assert.notCalled(callback);
+      instance._maybeSendPaintedEvent();
+
+      assert.isFalse(instance._timestampHandled);
+    });
+  });
 
+  describe("#_afterFramePaint", () => {
+    it("should call callback after the requestAnimationFrame callback returns", done => {
+      this.callback = () => done();
+      sandbox.spy(this, "callback");
+      const wrapper = shallow(<TopSitesPerfTimer {...DEFAULT_PROPS} />);
+      const instance = wrapper.instance();
+
+      instance._afterFramePaint(this.callback);
+
+      assert.notCalled(this.callback);
       mockRaf.step({count: 1});
-      assert.calledOnce(callback);
```

Comment:
This was a little confusing to figure out as there's no explicit check that the callback was called except that if it wasn't called, the test would have timed out.